drivers/gpio/gpio_pca95xx.c

```diff
@@ -0,0 +1,671 @@
+	    || (io_flags == (GPIO_INPUT | GPIO_OUTPUT))) {
+		return -ENOTSUP;
+	}
+
+	/* Open-drain support is per port, not per pin.
+	 * So can't really support the API as-is.
+	 */
+	if ((flags & GPIO_SINGLE_ENDED) != 0U) {
+		return -ENOTSUP;
+	}
+
+	ret = setup_pin_dir(dev, access_op, pin, flags);
+	if (ret) {
+		LOG_ERR("PCA95XX[0x%X]: error setting pin direction (%d)",
+			i2c_addr, ret);
+		goto done;
+	}
+
+	ret = setup_pin_pullupdown(dev, access_op, pin, flags);
+	if (ret) {
+		LOG_ERR("PCA95XX[0x%X]: error setting pin pull up/down "
+			"(%d)", i2c_addr, ret);
+		goto done;
+	}
+
+done:
+	return ret;
+}
+
+/**
+ * @brief Set the pin or port output
+ *
+ * @param dev Device struct of the PCA95XX
+ * @param access_op Access operation (pin or port)
+ * @param pin The pin number
+ * @param value Value to set (0 or 1)
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int gpio_pca95xx_write(struct device *dev, int access_op,
+				u32_t pin, u32_t value)
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	union gpio_pca95xx_port_data *port = &drv_data->reg_cache.output;
+	int ret;
+
+	if (!has_i2c_master(dev)) {
+		return -EINVAL;
+	}
+
+	/* Invert input value for pins configurated as active low. */
+	switch (access_op) {
+	case GPIO_ACCESS_BY_PIN:
+		if (value) {
+			port->all |= BIT(pin);
+		} else {
+			port->all &= ~BIT(pin);
+		}
+		break;
+	case GPIO_ACCESS_BY_PORT:
+		port->all = value;
+		break;
+	default:
+		ret = -ENOTSUP;
+		goto done;
+	}
+
+	ret = write_port_regs(dev, REG_OUTPUT_PORT0, port);
+
+done:
+	return ret;
+}
+
+/**
+ * @brief Read the pin or port status
+ *
+ * @param dev Device struct of the PCA95XX
+ * @param access_op Access operation (pin or port)
+ * @param pin The pin number
+ * @param value Value of input pin(s)
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int gpio_pca95xx_read(struct device *dev, int access_op,
+			       u32_t pin, u32_t *value)
+{
+	union gpio_pca95xx_port_data buf;
+	int ret;
+
+	if (!has_i2c_master(dev)) {
+		return -EINVAL;
+	}
+
+	ret = read_port_regs(dev, REG_INPUT_PORT0, &buf);
+	if (ret != 0) {
+		goto done;
+	}
+
+	switch (access_op) {
+	case GPIO_ACCESS_BY_PIN:
+		*value = (buf.all >> pin) & 0x01;
+		break;
+	case GPIO_ACCESS_BY_PORT:
+		*value = buf.all;
+		break;
+	default:
+		ret = -ENOTSUP;
+		break;
+	}
+
+done:
+	return ret;
+}
+
+static int gpio_pca95xx_port_get_raw(struct device *dev, u32_t *value)
+{
+	return gpio_pca95xx_read(dev, GPIO_ACCESS_BY_PORT, 0, value);
+}
+
+static int gpio_pca95xx_port_set_masked_raw(struct device *dev,
+					      u32_t mask, u32_t value)
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	union gpio_pca95xx_port_data *port = &drv_data->reg_cache.output;
+	u32_t new_value;
+
+	new_value = (port->all & ~mask) | (mask & value);
+
+	return gpio_pca95xx_write(dev, GPIO_ACCESS_BY_PORT, 0, new_value);
```

Comment:
The `GPIO_ACCESS_BY_PORT` will be soon removed from the API and - as mentioned by @pabigot - it's already unused. All the new `gpio_pca95xx_port_*` functions should use `write_port_regs`, `read_port_regs` directly.

Reply:
Does this mean that the current `.read()` and `.write()` APIs are going away, and that read/write by pin will no longer be supported?

Reply:
Yes, the `.read()` and `.write()` functions are going away. The API calls will be deprecated in this release and removed later. We want to remove their implementations already. The API will call new functions instead in a way that should be transparent to the user.

Reply:
In that case, I will duplicate the current read/write functions without the port access. So when it is time to remove the API, those old functions can be trivially removed.

Reply:
I think it will be cleaner if you call here `write_port_regs` directly. Only the mutex logic needs to be moved inside `write_port_regs` as well.

Reply:
We need also to update the cached value of the output register. Doing it in a way that's thread safe is a bit tricky. You may want to check `drivers/gpio/gpio_sx1509b.c` driver that has already been updated.

Reply:
I have updated the code to do this. So when GPIO_ACCESS_BY_PORT is remove, those functions can be removed trivially.